File: src/android/unity_ads/src/main/java/com/ee/UnityAdsBridge.kt
package com.ee

import android.app.Activity
import android.app.Application
import androidx.annotation.AnyThread
import com.ee.internal.UnityBannerAd
import com.ee.internal.UnityBannerHelper
import com.ee.internal.deserialize
import com.ee.internal.serialize
import com.unity3d.ads.IUnityAdsInitializationListener
import com.unity3d.ads.IUnityAdsLoadListener
import com.unity3d.ads.IUnityAdsShowListener
import com.unity3d.ads.UnityAds
import com.unity3d.ads.UnityAds.UnityAdsShowCompletionState
import com.unity3d.services.banners.UnityBannerSize
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.serialization.Serializable
import java.util.Collections
import java.util.concurrent.ConcurrentHashMap
import kotlin.coroutines.resume

class UnityAdsBridge(
    private val _bridge: IMessageBridge,
    private val _logger: ILogger,
    private val _application: Application,
    private var _activity: Activity?
) : IPlugin {
    companion object {
        private val kTag = UnityAdsBridge::class.java.name
        private const val kPrefix = "UnityAdsBridge"
        private const val kInitialize = "${kPrefix}Initialize"
        private const val kSetDebugModeEnabled = "${kPrefix}SetDebugModeEnabled"
        private const val kGetBannerAdSize = "${kPrefix}GetBannerAdSize"
        private const val kCreateBannerAd = "${kPrefix}CreateBannerAd"
        private const val kDestroyAd = "${kPrefix}DestroyAd"
        private const val kHasRewardedAd = "${kPrefix}HasRewardedAd"
        private const val kLoadRewardedAd = "${kPrefix}LoadRewardedAd"
        private const val kShowRewardedAd = "${kPrefix}ShowRewardedAd"
        private const val kOnLoaded = "${kPrefix}OnLoaded"
        private const val kOnFailedToShow = "${kPrefix}OnFailedToShow"
        private const val kOnClosed = "${kPrefix}OnClosed"
    }

    private var _initializing = false
    private var _initialized = false
    private val _bannerHelper = UnityBannerHelper()
    private val _ads: MutableMap<String, IAd> = ConcurrentHashMap()
    private val _loadedAdIds: MutableSet<String> = Collections.newSetFromMap(ConcurrentHashMap())

    init {
        _logger.info("$kTag: constructor begin: application = $_application activity = $_activity")
        registerHandlers()
        _logger.info("$kTag: constructor end")
    }

    override fun onCreate(activity: Activity) {
        _activity = activity
        for (ad in _ads.values) {
            ad.onCreate(activity)
        }
    }

    override fun onStart() {}
    override fun onStop() {}
    override fun onResume() {
        for (ad in _ads.values) {
            ad.onResume()
        }
    }

    override fun onPause() {
        for (ad in _ads.values) {
            ad.onPause()
        }
    }

    override fun onDestroy() {
        for (ad in _ads.values) {
            ad.onDestroy()
        }
        _activity = null
    }

    override fun destroy() {
        deregisterHandlers()
        for (ad in _ads.values) {
            ad.destroy()
        }
        _ads.clear()
    }

    @Serializable
    private class InitializeRequest(
        val gameId: String,
        val testModeEnabled: Boolean
    )

    @Serializable
    @Suppress("unused")
    private class GetBannerAdSizeResponse(
        val width: Int,
        val height: Int
    )

    @Serializable
    private class CreateBannerAdRequest(
        val adId: String,
        val adSize: Int
    )

    @AnyThread
    private fun registerHandlers() {
        _bridge.registerAsyncHandler(kInitialize) { message ->
            val request = deserialize<InitializeRequest>(message)
            Utils.toString(initialize(request.gameId, request.testModeEnabled))
        }
        _bridge.registerHandler(kSetDebugModeEnabled) { message ->
            setDebugModeEnabled(Utils.toBoolean(message))
            ""
        }
        _bridge.registerHandler(kGetBannerAdSize) { message ->
            val index = message.toInt()
            val size = _bannerHelper.getSize(index)
            val response = GetBannerAdSizeResponse(size.x, size.y)
            response.serialize()
        }
        _bridge.registerHandler(kCreateBannerAd) { message ->
            val request = deserialize<CreateBannerAdRequest>(message)
            val adSize = _bannerHelper.getAdSize(request.adSize)
            Utils.toString(createBannerAd(request.adId, adSize))
        }
        _bridge.registerHandler(kDestroyAd) { message ->
            Utils.toString(destroyAd(message))
        }
        _bridge.registerHandler(kHasRewardedAd) { message ->
            Utils.toString(hasRewardedAd(message))
        }
        _bridge.registerAsyncHandler(kLoadRewardedAd) { message ->
            Utils.toString(loadRewardedAd(message))
        }
        _bridge.registerHandler(kShowRewardedAd) { message ->
            showRewardedAd(message)
            ""
        }
    }

    @AnyThread
    private fun deregisterHandlers() {
        _bridge.deregisterHandler(kInitialize)
        _bridge.deregisterHandler(kSetDebugModeEnabled)
        _bridge.deregisterHandler(kGetBannerAdSize)
        _bridge.deregisterHandler(kCreateBannerAd)
        _bridge.deregisterHandler(kDestroyAd)
        _bridge.deregisterHandler(kHasRewardedAd)
        _bridge.deregisterHandler(kLoadRewardedAd)
        _bridge.deregisterHandler(kShowRewardedAd)
    }

    @AnyThread
    suspend fun initialize(gameId: String, testModeEnabled: Boolean): Boolean {
        return suspendCancellableCoroutine { cont ->
            Thread.runOnMainThread {
                if (!UnityAds.isSupported()) {
                    _logger.info("$kTag: initialize: not supported")
                    cont.resume(false)
                    return@runOnMainThread
                }
                if (UnityAds.isInitialized()) {
                    _initialized = true
                    _logger.info("$kTag: initialize: initialized")
                    cont.resume(true)
                    return@runOnMainThread
                }
                if (_initializing) {
                    _logger.info("$kTag: initialize: initializing")
                    cont.resume(false)
                    return@runOnMainThread
                }
                if (_initialized) {
                    _logger.info("$kTag: initialize: initialized")
                    cont.resume(true)
                    return@runOnMainThread
                }
                _initializing = true
                UnityAds.initialize(
                    _application,
                    gameId,
                    testModeEnabled,
                    object : IUnityAdsInitializationListener {
                        override fun onInitializationComplete() {
                            if (cont.isActive) {
                                // OK.
                            } else {
                                return
                            }
                            Thread.runOnMainThread {
                                _logger.info("$kTag: initialize: done")
                                _initializing = false
                                _initialized = true
                                cont.resume(true)
                            }
                        }

                        override fun onInitializationFailed(
                            error: UnityAds.UnityAdsInitializationError?,
                            message: String?
                        ) {
                            if (cont.isActive) {
                                // OK.
                            } else {
                                return
                            }
                            Thread.runOnMainThread {
                                _logger.error("${this::onInitializationFailed.name}: error = ${error ?: ""} message = ${message ?: ""}")
                                _initializing = false
                                _initialized = true
                                cont.resume(true)
                            }
                        }
                    })
            }
        }
    }

    @AnyThread
    private fun setDebugModeEnabled(enabled: Boolean) {
        Thread.runOnMainThread {
            if (!_initialized) {
                return@runOnMainThread
            }
            UnityAds.setDebugMode(enabled)
        }
    }

    @AnyThread
    fun createBannerAd(adId: String, adSize: UnityBannerSize): Boolean {
        return createAd(adId) {
            UnityBannerAd(_bridge, _logger, _activity, adId, adSize, _bannerHelper)
        }
    }

    @AnyThread
    fun createAd(adId: String, creator: () -> IAd): Boolean {
        if (_ads.containsKey(adId)) {
            return false
        }
        val ad = creator()
        _ads[adId] = ad
        return true
    }

    @AnyThread
    fun destroyAd(adId: String): Boolean {
        val ad = _ads[adId] ?: return false
        ad.destroy()
        _ads.remove(adId)
        return true
    }

    @AnyThread
    fun hasRewardedAd(adId: String): Boolean {
        if (!_initialized) {
            // Changed on main thread.
            return false
        }
        return _loadedAdIds.contains(adId)
    }

    @AnyThread
    suspend fun loadRewardedAd(adId: String): Boolean {
        return suspendCancellableCoroutine { cont ->
            Thread.runOnMainThread {
                _logger.debug("$kTag: loadRewardedAd: id = $adId")
                if (!_initialized) {
                    _logger.error("$kTag: loadRewardedAd: not initialized")
                    cont.resume(false)
                    return@runOnMainThread
                }
                UnityAds.load(adId, object : IUnityAdsLoadListener {
                    override fun onUnityAdsAdLoaded(placementId: String) {
                        if (cont.isActive) {
                            // OK.
                        } else {
                            return
                        }
                        Thread.runOnMainThread {
                            _logger.debug("$kTag: ${this::onUnityAdsAdLoaded.name}: id = $adId")
                            _loadedAdIds.add(adId)
                            cont.resume(true)
                        }
                    }

                    override fun onUnityAdsFailedToLoad(
                        placementId: String,
                        error: UnityAds.UnityAdsLoadError,
                        message: String
                    ) {
                        if (cont.isActive) {
                            // OK.
                        } else {
                            return
                        }
                        Thread.runOnMainThread {
                            _logger.debug("$kTag: ${this::onUnityAdsFailedToLoad.name}: id = $adId")
                            cont.resume(false)
                        }
                    }
                })
            }
        }
    }

    @Serializable
    @Suppress("unused")
    private class ErrorResponse(
        val ad_id: String,
        val message: String
    )

    @Serializable
    @Suppress("unused")
    private class ResultResponse(
        val ad_id: String,
        val rewarded: Boolean
    )

    @AnyThread
    fun showRewardedAd(adId: String) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::showRewardedAd.name}: id = $adId")
            if (!_initialized) {
                _logger.error("$kTag: ${this::showRewardedAd.name}: not initialized")
                val response = ErrorResponse(adId, "not initialized")
                _bridge.callCpp(kOnFailedToShow, response.serialize())
                return@runOnMainThread
            }
            UnityAds.show(_activity, adId, object : IUnityAdsShowListener {
                override fun onUnityAdsShowStart(placementId: String?) {
                    _logger.debug("$kTag: ${this::onUnityAdsShowStart.name}: placementId = $placementId")
                }

                override fun onUnityAdsShowClick(placementId: String?) {
                    _logger.debug("$kTag: ${this::onUnityAdsShowClick.name}: placementId = $placementId")
                }

                override fun onUnityAdsShowComplete(
                    placementId: String?,
                    state: UnityAdsShowCompletionState?
                ) {
                    Thread.runOnMainThread {
                        _logger.debug("$kTag: ${this::onUnityAdsShowComplete.name}: placementId = $placementId state = $state")
                        _loadedAdIds.remove(placementId)
                        val response =
                            ResultResponse(adId, state == UnityAdsShowCompletionState.COMPLETED)
                        _bridge.callCpp(kOnClosed, response.serialize())
                    }
                }

                override fun onUnityAdsShowFailure(
                    placementId: String?,
                    error: UnityAds.UnityAdsShowError?,
                    message: String?
                ) {
                    Thread.runOnMainThread {
                        _logger.debug("$kTag: ${this::onUnityAdsShowFailure.name}: placementId = $placementId message = $message")
                        _loadedAdIds.remove(placementId)
                        val response = ErrorResponse(adId, "")
                        _bridge.callCpp(kOnFailedToShow, response.serialize())
                    }
                }
            })
        }
    }
}

File: src/ios/ee/unity_ads/UnityAdsBridge.swift
//
//  UnityAds.swift
//  ee-x-366f64ba
//
//  Created by eps on 6/25/20.
//

import RxSwift
import UnityAds

private let kTag = "\(UnityAdsBridge.self)"
private let kPrefix = "UnityAdsBridge"
private let kInitialize = "\(kPrefix)Initialize"
private let kSetDebugModeEnabled = "\(kPrefix)SetDebugModeEnabled"
private let kGetBannerAdSize = "\(kPrefix)GetBannerAdSize"
private let kCreateBannerAd = "\(kPrefix)CreateBannerAd"
private let kDestroyAd = "\(kPrefix)DestroyAd"
private let kHasRewardedAd = "\(kPrefix)HasRewardedAd"
private let kLoadRewardedAd = "\(kPrefix)LoadRewardedAd"
private let kShowRewardedAd = "\(kPrefix)ShowRewardedAd"
private let kOnLoaded = "\(kPrefix)OnLoaded"
private let kOnFailedToShow = "\(kPrefix)OnFailedToShow"
private let kOnClosed = "\(kPrefix)OnClosed"

private class InitializeDelegate: NSObject, UnityAdsInitializationDelegate {
    private let _onCompleted: () -> Void
    private let _onFailed: (UnityAdsInitializationError, String) -> Void

    public init(_ onCompleted: @escaping () -> Void,
                _ onFailed: @escaping (UnityAdsInitializationError, String) -> Void) {
        _onCompleted = onCompleted
        _onFailed = onFailed
    }

    func initializationComplete() {
        _onCompleted()
    }

    func initializationFailed(_ error: UnityAdsInitializationError, withMessage message: String) {
        _onFailed(error, message)
    }
}

private class LoadDelegate: NSObject, UnityAdsLoadDelegate {
    private let _onLoaded: (String) -> Void
    private let _onFailed: (String) -> Void

    public init(_ onLoaded: @escaping (String) -> Void,
                _ onFailed: @escaping (String) -> Void) {
        _onLoaded = onLoaded
        _onFailed = onFailed
    }

    func unityAdsAdLoaded(_ placementId: String) {
        _onLoaded(placementId)
    }

    func unityAdsAdFailed(toLoad placementId: String, withError error: UnityAdsLoadError, withMessage message: String) {
        _onFailed(placementId)
    }
}

private class ShowDelegate: NSObject, UnityAdsShowDelegate {
    private let _onFailedToShow: (String) -> Void
    private let _onClosed: (Bool) -> Void
    
    public init(_ onFailedToShow: @escaping (String) -> Void,
                _ onClosed: @escaping (Bool) -> Void) {
        _onFailedToShow = onFailedToShow
        _onClosed = onClosed
    }
    
    func unityAdsShowComplete(_ placementId: String, withFinish state: UnityAdsShowCompletionState) {
        if (state == UnityAdsShowCompletionState.showCompletionStateCompleted) {
            _onClosed(true)
        } else {
            _onClosed(false)
        }
    }
    
    func unityAdsShowFailed(_ placementId: String, withError error: UnityAdsShowError, withMessage message: String) {
        _onFailedToShow(message)
    }
    
    func unityAdsShowStart(_ placementId: String) {
        // TODO.
    }
    
    func unityAdsShowClick(_ placementId: String) {
        // TODO.
    }
}

@objc(EEUnityAdsBridge)
class UnityAdsBridge: NSObject, IPlugin {
    private let _bridge: IMessageBridge
    private let _logger: ILogger
    private var _initializing = false
    private var _initialized = false
    private let _bannerHelper = UnityBannerHelper()
    private var _ads: [String: IAd] = [:]
    private var _displayed = false
    private var _loadedAdIds: Set<String> = []

    public required init(_ bridge: IMessageBridge, _ logger: ILogger) {
        _bridge = bridge
        _logger = logger
        super.init()
        registerHandlers()
    }

    public func destroy() {
        deregisterHandlers()
        _ads.values.forEach { $0.destroy() }
        _ads.removeAll()
    }

    func registerHandlers() {
        _bridge.registerAsyncHandler(kInitialize) { message, resolver in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard
                let gameId = dict["gameId"] as? String,
                let testModeEnabled = dict["testModeEnabled"] as? Bool
            else {
                assert(false, "Invalid argument")
                return
            }
            self.initialize(gameId, testModeEnabled)
                .subscribe(
                    onSuccess: {
                        result in resolver(Utils.toString(result))
                    }, onError: {
                        _ in resolver(Utils.toString(false))
                    })
        }
        _bridge.registerHandler(kSetDebugModeEnabled) { message in
            self.setDebugModeEnabled(Utils.toBool(message))
            return ""
        }
        _bridge.registerHandler(kGetBannerAdSize) { message in
            let index = Int(message) ?? -1
            let size = self._bannerHelper.getSize(index: index)
            return EEJsonUtils.convertDictionary(toString: [
                "width": size.width,
                "height": size.height
            ])
        }
        _bridge.registerHandler(kCreateBannerAd) { message in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard
                let adId = dict["adId"] as? String,
                let index = dict["adSize"] as? Int
            else {
                assert(false, "Invalid argument")
                return ""
            }
            let adSize = self._bannerHelper.getAdSize(index)
            return Utils.toString(self.createBannerAd(adId, adSize))
        }
        _bridge.registerHandler(kDestroyAd) { message in
            Utils.toString(self.destroyAd(message))
        }
        _bridge.registerHandler(kHasRewardedAd) { message in
            Utils.toString(self.hasRewardedAd(message))
        }
        _bridge.registerAsyncHandler(kLoadRewardedAd) { message, resolver in
            self.loadRewardedAd(message)
                .subscribe(
                    onSuccess: {
                        result in resolver(Utils.toString(result))
                    }, onError: {
                        _ in resolver(Utils.toString(false))
                    })
        }
        _bridge.registerHandler(kShowRewardedAd) { message in
            self.showRewardedAd(message)
            return ""
        }
    }

    func deregisterHandlers() {
        _bridge.deregisterHandler(kInitialize)
        _bridge.deregisterHandler(kSetDebugModeEnabled)
        _bridge.deregisterHandler(kGetBannerAdSize)
        _bridge.deregisterHandler(kCreateBannerAd)
        _bridge.deregisterHandler(kDestroyAd)
        _bridge.deregisterHandler(kHasRewardedAd)
        _bridge.deregisterHandler(kLoadRewardedAd)
        _bridge.deregisterHandler(kShowRewardedAd)
    }

    func initialize(_ gameId: String, _ testModeEnabled: Bool) -> Single<Bool> {
        return Single<Bool>.create { single in
            Thread.runOnMainThread {
                if !UnityAds.isSupported() {
                    self._logger.info("\(kTag): \(#function): not supported")
                    single(.success(false))
                    return
                }
                if UnityAds.isInitialized() {
                    self._initialized = true
                    self._logger.info("\(kTag): \(#function): initialized")
                    single(.success(true))
                    return
                }
                if self._initializing {
                    self._logger.info("\(kTag): \(#function): initializing")
                    single(.success(false))
                    return
                }
                if self._initialized {
                    self._logger.info("\(kTag): \(#function): initialized")
                    single(.success(true))
                    return
                }
                self._initializing = true
                UnityAds.initialize(gameId, testMode: testModeEnabled, initializationDelegate: InitializeDelegate {
                    Thread.runOnMainThread {
                        self._logger.debug("\(kTag): initializationComplete")
                        self._initializing = false
                        self._initialized = true
                        single(.success(true))
                    }
                } _: { error, message in
                    Thread.runOnMainThread {
                        self._logger.debug("\(kTag): initializationFailed: error = \(error) message = \(message)")
                        self._initializing = false
                        single(.success(false))
                    }
                 })
            }
            return Disposables.create()
        }
        .subscribeOn(MainScheduler())
    }

    func setDebugModeEnabled(_ enabled: Bool) {
        Thread.runOnMainThread {
            if !self._initialized {
                return
            }
            UnityAds.setDebugMode(enabled)
        }
    }

    func createBannerAd(_ adId: String, _ adSize: CGSize) -> Bool {
        return createAd(adId) {
            UnityBannerAd(self._bridge, self._logger, adId, adSize, self._bannerHelper)
        }
    }

    func createAd(_ adId: String, _ creator: () -> IAd) -> Bool {
        if _ads.contains(where: { key, _ in key == adId }) {
            return false
        }
        let ad = creator()
        _ads[adId] = ad
        return true
    }

    func destroyAd(_ adId: String) -> Bool {
        guard let ad = _ads[adId] else {
            return false
        }
        ad.destroy()
        _ads.removeValue(forKey: adId)
        return true
    }

    func hasRewardedAd(_ adId: String) -> Bool {
        if !_initialized {
            return false
        }
        return _loadedAdIds.contains(adId)
    }

    func loadRewardedAd(_ adId: String) -> Single<Bool> {
        return Single<Bool>.create { single in
            Thread.runOnMainThread {
                self._logger.debug("\(kTag): \(#function): id = \(adId)")
                if !self._initialized {
                    self._logger.debug("\(kTag): \(#function): not initialized")
                    single(.success(false))
                    return
                }
                UnityAds.load(adId, loadDelegate: LoadDelegate { _ in
                    Thread.runOnMainThread {
                        self._logger.debug("\(kTag): unityAdsAdLoaded: \(adId)")
                        single(.success(true))
                    }
                } _: { _ in
                    Thread.runOnMainThread {
                        self._logger.debug("\(kTag): unityAdsAdFailed: \(adId)")
                        single(.success(false))
                    }
                 })
            }
            return Disposables.create()
        }
        .subscribeOn(MainScheduler())
    }

    func showRewardedAd(_ adId: String) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function): id = \(adId)")
            if !self._initialized {
                self._logger.debug("\(kTag): \(#function): not initialized")
                self._bridge.callCpp(kOnFailedToShow, EEJsonUtils.convertDictionary(toString: [
                    "ad_id": adId,
                    "message": ""
                ]))
                return
            }
            guard let rootView = Utils.getCurrentRootViewController() else {
                assert(false, "Root view is null")
                return
            }
            UnityAds.show(rootView, placementId: adId, showDelegate: ShowDelegate { message in
                Thread.runOnMainThread {
                    self._logger.debug("\(kTag): unityAdsOnFailedToShow: \(adId)")
                    self._bridge.callCpp(kOnFailedToShow, EEJsonUtils.convertDictionary(toString: [
                        "ad_id": adId,
                        "message": message
                    ]))
                }
            } _: { rewarded in
                Thread.runOnMainThread {
                    self._logger.debug("\(kTag): unityAdsOnClosed: \(adId) rewarded = \(rewarded)")
                    self._bridge.callCpp(kOnClosed, EEJsonUtils.convertDictionary(toString: [
                        "ad_id": adId,
                        "rewarded": rewarded
                    ]))
                }
            })
        }
    }

    func unityAdsReady(_ adId: String) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function): id = \(adId)")
            self._loadedAdIds.insert(adId)
            self._bridge.callCpp(kOnLoaded, adId)
        }
    }

    func unityAdsDidStart(_ adId: String) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function): id = \(adId)")
        }
    }
}
